src/hotspot/share/memory/metaspaceShared.cpp

```diff
@@ -592,10 +593,11 @@ void MetaspaceShared::link_and_cleanup_shared_classes(TRAPS) {
     // Keep scanning until we have linked no more classes.
   }
 
-  for (int i = 0; i < _loaded_cld->length(); i++) {
-    ClassLoaderData* cld = _loaded_cld->at(i);
+  for (int i = 0; i < loaded_cld.length(); i++) {
+    ClassLoaderData* cld = loaded_cld.at(i);
     cld->dec_keep_alive();
   }
+  loaded_cld.trunc_to(0);
```

Comment:
There's no need for the trucate -- `loaded_cld` is locally allocated and will be freed after this function returns.

Also, to improve modularity, I think we should move the dec_keep_alive loop into the destructor of CollectCLDClosure.

Also, `loaded_cld` can be moved as a field into CollectCLDClosure.